draft-ietf-jsonpath-base.md

```diff
@@ -1006,55 +1005,66 @@ The current item is selected if and only if the result is `true`.
 
 
 ~~~~ abnf
-boolean-expr     = logical-or-expr
-logical-or-expr  = logical-and-expr *(S "||" S logical-and-expr)
-                                                      ; disjunction
-                                                      ; binds less tightly than conjunction
-logical-and-expr = basic-expr *(S "&&" S basic-expr)  ; conjunction
-                                                      ; binds more tightly than disjunction
+boolean-expr = or-expr   
+
+or-expr      = and-expr *(S "||" S and-expr)        ; logical or operator
+
+and-expr     = basic-expr *(S "&&" S basic-expr)    ; logical and operator
 
 basic-expr   = exist-expr /
                paren-expr /
                relation-expr
-exist-expr   = [neg-op S] path                          ; path existence or non-existence
+exist-expr   = [neg-op S] path                      ; path existence or non-existence
 path         = rel-path / json-path
 rel-path     = "@" *(S (dot-selector / index-selector))
-paren-expr   = [neg-op S] "(" S boolean-expr S ")"    ; parenthesized expression
-neg-op       = "!"                                    ; not operator
+paren-expr   = [neg-op S] "(" S boolean-expr S ")"  ; parenthesized expression
+neg-op       = "!"                                  ; not operator
 
-relation-expr = comp-expr /                           ; comparison test
-                regex-expr                            ; regular expression test
+relation-expr = comp-expr /                         ; comparison test
+                regex-expr                          ; regular expression test
 
 comp-expr    = comparable S comp-op S comparable
-comparable   = number / string-literal /              ; primitive ...
-               true / false / null /                  ; values only
-               path                                   ; path value
-comp-op      = "==" / "!=" /                          ; comparison ...
-               "<"  / ">"  /                          ; operators
-               "<=" / ">="
-number       = int [ frac ] [ exp ]                   ; decimal number
-frac         = "." 1*DIGIT                            ; decimal fraction
-exp          = "e" [ "-" / "+" ] 1*DIGIT              ; decimal exponent
-true         = %x74.72.75.65                          ; true
-false        = %x66.61.6c.73.65                       ; false
-null         = %x6e.75.6c.6c                          ; null
+comparable   = json-value / path                    ; JSON value or path value
+json-value   = number / string-literal /            ; primitive
```

Comment:
To me it seemed the resolution of #127 at https://datatracker.ietf.org/doc/minutes-112-jsonpath/ implied that we weren't going to entertain notating structured values in a JSONPath expression at all.  So this change surprises me a little.
Maybe this was just my perception, or if not, maybe we want to revert that aspect.

Reply:
hmm ... I am quite sure that we somewhere, somehow agreed, that excluding structured values from equality ...
```
color == [12, 125, 255]
```
is kind of arbitrary and artificial.  Can't find it yet though. Glyn for help here ... `?

Reply:
> We also have to define the semantics, which is relatively straightforward for most kinds of JSON values (lexicographic ordering). 

You're not suggesting lexicographic ordering for numbers, are you?

> But there is no natural total ordering for maps that I am aware of, so what does it mean to have a `<` in an expression (which, of course, needs to stay syntactically allowed) and then two maps turning up? Oh, and what is the value of `1 < true` (already a problem without this extension)? What does it mean that `<` does not "work"?

All of these issues have been addressed in prior work, in different ways. JMESPath, which defines ordering operators only for numbers, specifies exactly what it means for `1 < true` or `1 < {}`, see [ordering operators](https://jmespath.org/specification.html#ordering-operators). Other approaches define a complete ordering on JSON values, assuming where types differ that null < boolean < number < object < array < string (following Python.)

Reply:
I am proposing:
```
* Relational operators (`>`, `<`, `>=`, `<=`) are intended for comparing primitive number values. 
Comparison with a non-number value(s) fails, i.e. no selection occurs.
``` 
So there is no need to discuss ordering of structured values. In that case no selection occurs,
We can include strings on basis of lexical ordering though.

Regarding `1 < true` ... we opposed against implicit type conversion, so no selection occurs here.

But ... as I was always in favor of comparing primitive values only, I would be happy to return to that clean solution again. I am obviously wrong in interpreting the discussion in context `contain-expr` (which is obsolete now) that way, that it is natural to include structured values in comparisons. (#127)

Reply:
My recollection, which could easily be faulty, is that we didn't want to support non-primitive literal values because of the cost/benefit ratio being high. I *think* the jury was out on whether we would support ("deep") equals comparisons of non-primitive (non-literal) values, but again I may have misremembered. It's hard to keep track of past discussions...

Reply:
> My recollection, which could easily be faulty, is that we didn't want to support non-primitive literal values because of the cost/benefit ratio being high. I _think_ the jury was out on whether we would support ("deep") equals comparisons of non-primitive (non-literal) values, but again I may have misremembered. It's hard to keep track of past discussions...

On the other hand, there's plenty of prior experience to draw on, including a few JSONPath implementations, and the very completely specified JMESPath. It's not necessary to think this all through as if from the beginning.